src/v/cluster/controller_backend.cc

```diff
@@ -231,14 +231,48 @@ std::vector<topic_table::delta> calculate_bootstrap_deltas(
     std::move(start, deltas.end(), std::back_inserter(result_delta));
     return result_delta;
 }
-
+bool is_cross_core_update(model::node_id self, const topic_table_delta& delta) {
+    if (!delta.previous_assignment) {
+        return false;
+    }
+    return contains_node(self, delta.previous_assignment->replicas)
+           && !has_local_replicas(self, delta.previous_assignment->replicas);
+}
 ss::future<>
 controller_backend::bootstrap_ntp(const model::ntp& ntp, deltas_t& deltas) {
     vlog(clusterlog.trace, "bootstrapping {}", ntp);
     // find last delta that has to be applied
-    auto bootstrap_deltas = calculate_bootstrap_deltas(
-      _self, std::move(deltas));
+    auto bootstrap_deltas = calculate_bootstrap_deltas(_self, deltas);
+    auto& first_delta = bootstrap_deltas.front();
+    // if first operation is a cross core update, find initial revision on
+    // current node and store it in bootstrap map
+    using op_t = topic_table::delta::op_type;
+    if (is_cross_core_update(_self, first_delta)) {
+        // find opeartion that created current partition on this node
+        auto it = std::find_if(
+          deltas.rbegin(),
+          deltas.rend(),
+          [this](const topic_table_delta& delta) {
+              if (delta.type == op_t::add) {
+                  return true;
+              }
+              return delta.type == op_t::update_finished
+                     && !contains_node(_self, delta.new_assignment.replicas);
+          });
 
+        vassert(
+          it != deltas.rend(),
+          "if partition {} was moved from different core it had to exists on "
+          "current node previously",
+          ntp);
+        // if we found update finished operation it is preceeding operation that
+        // created partition on current node
+        if (it->type == op_t::update_finished) {
+            it = std::prev(it);
```

Comment:
I think it is impossible to `it` to be `rend()++`. `rend()` is already not dereferencable since it points to element before first one in vector. 
From cppreference:
```
For a reverse iterator r constructed from an iterator i, the relationship &*r == &*(i-1) is always true (as long as r is dereferenceable); thus a reverse iterator constructed from a one-past-the-end iterator dereferences to the last element in a sequence.
```
There is also nice diagram explaining this: [reverse iterator](https://en.cppreference.com/w/cpp/iterator/reverse_iterator)

I've added an assertion checking that when delta is of type `update_finished` it has to have preceding operation:
```c++
  if (it->type == op_t::update_finished) {
            vassert(
              it != deltas.rbegin(),
              "update finished delta {} must have preceeding operation",
              *it);
            it = std::prev(it);
        }
```

Reply:
ahh, thanks I was confused about the directions. the assertion you added is great, that's what I had in mind.